source/common/network/io_socket_handle_impl.cc

```diff
@@ -86,6 +88,22 @@ Api::IoCallUint64Result IoSocketHandleImpl::readv(uint64_t max_length, Buffer::R
   absl::FixedArray<iovec> iov(num_slice);
   uint64_t num_slices_to_read = 0;
   uint64_t num_bytes_to_read = 0;
+
+  if (injected_data_ != nullptr && injected_data_->length() != 0) {
+    for (; num_slices_to_read < num_slice && num_bytes_to_read < max_length; num_slices_to_read++) {
+      auto size_to_copy = std::min(slices[num_slices_to_read].len_, max_length - num_bytes_to_read);
+      size_to_copy = std::min(size_to_copy, injected_data_->length());
+      injected_data_->copyOut(0, size_to_copy, slices[num_slices_to_read].mem_);
```

Comment:
We will have data copy here. But I guess it should be ok, since the real socket have data copy also.